3rd_party_slots/python_client/hyperon_das/service_clients/atomdb.py

```diff
@@ -34,69 +37,86 @@ def build_atoms_from_tokens(self, tokens: list[str]):
-            flush()
+            if len(buffer) != 0 and current:
+                flush()
 
-        return atoms
+            return atoms
 
     def add_atoms(self, atoms: list[Atom]) -> list[str]:
         """
-        Serialize atoms, send `ADD_ATOMS` to the remote peer, and return their handles.
-
-        Each atom must implement `tokenize(args: list[str])` and `handle() -> str`.
-        The method prepends the atom-type token ("NODE" / "LINK") to each atom's
-        token block before sending.
-
-        Args:
-            atoms: List of Atom instances to serialize and send.
-
-        Returns:
-            List of handles corresponding to the provided atoms, in the same order.
-
-        Raises:
-            ValueError: If an atom is neither `Node` nor `Link`.
+        Serialize atoms into args, split args into chunks if necessary to respect the
+        message size limit, and then send each chunk to the remote peer.
         """
-        args = []
-        handles = []
-        atom_type = ""
-
-        for atom in atoms:
-            atom.tokenize(args)
-
-            if isinstance(atom, Node):
-                atom_type = "NODE"
-            elif isinstance(atom, Link):
-                atom_type = "LINK"
-            else:
-                log.error("Invalid Atom Type")
-                raise ValueError("Invalid Atom Type")
-
-            args.insert(0, atom_type)
-            handles.append(atom.handle())
-
-        self.to_remote_peer(self.ADD_ATOMS, args)
-
-        return handles
+        with self._lock:
+            handles = []
+            chunks = []
+            current_chunk = []
+            current_size = 0
+            temp_atom_args = []
+            atoms_count = 0
+
+            for atom in atoms:
+                handles.append(atom.handle())
+
+                temp_atom_args.clear()
+                atom.tokenize(temp_atom_args)
+
+                if isinstance(atom, Node):
+                    atom_type = "NODE"
+                elif isinstance(atom, Link):
+                    atom_type = "LINK"
+                else:
+                    log.error(f"Invalid Atom Type: {type(atom)}")
+                    raise ValueError("Invalid Atom Type")
+
+                args_size = sum(len(arg.encode('utf-8')) for arg in ([atom_type] + temp_atom_args))
+
+                if current_size + args_size > MAX_ARGS_BYTES and current_chunk:
+                    chunks.append((current_chunk, atoms_count))
+                    current_chunk = []
+                    current_size = 0
+                    atoms_count = 0
+
+                current_chunk.extend([atom_type] + temp_atom_args)
+                atoms_count += 1
+                current_size += args_size
+
+            if current_chunk:
+                chunks.append((current_chunk, atoms_count))
+
+            log.debug(f"Sending {len(atoms)} atoms in {len(chunks)} chunk(s).")
+
+            for i, (chunk_args, atoms_count) in enumerate(chunks):
```

Comment:
@marcocapozzoli  You can merge the code this way but please add a `#FIXME TODO` here to fix this algorithm. Instead of building the entire chunks vector before start sending, you are supposed to send the chunks as soon as they are ready. This way you increase performance by allowing concurrent processing of the atoms in the server side and here (and probably you could get rid of the `sleep()` as well.